File: 20-valid-parentheses/20-valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
   
    const brackets = {
        "{" : "}", 
        "[" : "]",
        "(" : ")"
    }
    
    let closingBra = []
    
    for (let i = 0; i < s.length ; i++) {
        if (brackets.hasOwnProperty(s[i])) {
            closingBra.push(brackets[s[i]])
        } else {
            if (s[i] !== closingBra.pop()) return false
        }
    }
    return !closingBra.length 
};
    
    

File: 20-valid-parentheses/20-valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if len(s) % 2 != 0:
            return False
        dic = {
            '(' : ')',
            '[' : ']',
            '{' : '}'
        }
        stack = []
        
        for i , char in enumerate(s):
            if s[i] in dic:
                stack.append(char)
            else:
                if len(stack) != 0 and dic[stack[-1]] == char:
                    stack.pop()
                else:
                    return False
                
        return len(stack) == 0